### PR TITLE
Support schema 4 4

### PR DESCRIPTION
**Pull request by richardhallett**
## Purpose

closes: #104 

## Approach
Adds xsd resources / updates version 4
Adds support for resourceType mapping
Adds support for subject classification parsing.
Add supports for relatedItems - this is a new sub array containing various parts.

So it's clear here is currently how the output of the JSON is looking for the new relatedItems

```{
  "id": "https://doi.org/10.5072/example-full",
  "relatedItems": [
    {
      "relationType": "IsPublishedIn",
      "relatedItemType": "Journal",
      "relatedItemIdentifier": {
        "relatedItemIdentifier": "10.5072/john-smiths-1234",
        "relatedItemIdentifierType": "DOI",
        "relatedMetadataScheme": "citeproc+json",
        "schemeURI": "https://github.com/citation-style-language/schema/raw/master/csl-data.json",
        "schemeType": "URL"
      },
      "creators": [
        {
          "nameType": "Personal",
          "name": "Smith, John",
          "givenName": "John",
          "familyName": "Smith"
        }
      ],
      "titles": [
        {
          "title": "Understanding the fictional John Smith"
        },
        {
          "title": "A detailed look",
          "titleType": "Subtitle"
        }
      ],
      "volume": "776",
      "issue": "1",
      "number": "1",
      "firstPage": "50",
      "lastPage": "60",
      "publisher": "Example Inc",
      "publicationYear": "1776",
      "edition": "1",
      "contributors": [
        {
          "name": "Hallett, Richard",
          "givenName": "Richard",
          "familyName": "Hallett",
          "contributorType": "ProjectLeader"
        }
      ]
    }
  ],
}
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)

- [X] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Reviewer, please remember our [guidelines](https://datacite.atlassian.net/wiki/spaces/TEC/pages/1168375809/Pull+Request+Guidelines):

- Be humble in the language and feedback you give, ask don't tell.
- Consider using positive language as opposed to neutral when offering feedback. This is to avoid the negative bias that can occur with neutral language appearing negative.
- Offer suggestions on how to improve code e.g. simplification or expanding clarity.
- Ensure you give reasons for the changes you are proposing.

### CODE REVIEW

File: lib/bolognese/datacite_utils.rb [hunk 1]
```diff
@@ -209,6 +210,71 @@ def insert_related_identifiers(xml)
...
+              end
+            end
+
+            xml.volume(related_item['volume'])
+            xml.issue(related_item['issue'])
+            xml.number(related_item['number'], {'numberType' => related_item['numberType']}.compact)
+            xml.firstPage(related_item['firstPage'])
+            xml.lastPage(related_item['lastPage'])
+            xml.publisher(related_item['publisher'])
+            xml.publicationYear(related_item['publicationYear'])
```
Comment (mfenner) [inline lib/bolognese/datacite_utils.rb:262]: publicationYear should go before volume.

File: lib/bolognese/readers/datacite_reader.rb [hunk 2]
```diff
@@ -194,6 +186,42 @@ def read_datacite(string: nil, **options)
...
+            "creators" => get_authors(Array.wrap(ri.dig("creators", "creator"))),
+            "titles" => get_titles(ri),
+            "volume" => ri["volume"],
+            "issue" => ri["issue"],
+            "number" => ri.dig("number", "__content__"),
+            "numberType" => ri.dig("number", "numberType"),
+            "firstPage" => ri["firstPage"],
+            "lastPage" => ri["lastPage"],
+            "publisher" => ri["publisher"],
+            "publicationYear" => ri["publicationYear"],
```
Comment (mfenner) [inline lib/bolognese/readers/datacite_reader.rb:219]: publicationYear should go after titles and before volume.

File: lib/bolognese/utils.rb [hunk 3]
```diff
@@ -57,6 +57,19 @@ module Utils
       "Sound" => "AudioObject",
       "Text" => "ScholarlyArticle",
       "Workflow" => nil,
+      "Book" => "Book",
```
Comment (mfenner) [inline lib/bolognese/utils.rb:60]: Do we want to order all resourceTypeGeneral alphabetically? We do in the XSD.

File: lib/bolognese/utils.rb [hunk 4]
```diff
@@ -57,6 +57,19 @@ module Utils
...
+      "ConferencePaper" => "Article",
+      "ConferenceProceeding" => nil,
+      "Dissertation" => "Thesis",
+      "Journal" => nil,
+      "JournalArticle" => "ScholarlyArticle",
+      "OutputManagementPlan" => nil,
+      "PeerReview" => "Review",
+      "Preprint" => nil,
+      "Report" => "Report",
+      "Standard" => nil,
```
Comment (mfenner) [inline lib/bolognese/utils.rb:72]: We can map Journal amd ConferenceProceeding to Periodical.

File: lib/bolognese/readers/datacite_reader.rb [hunk 5]
```diff
@@ -92,15 +92,7 @@ def read_datacite(string: nil, **options)
           "ris" => Bolognese::Utils::CR_TO_RIS_TRANSLATIONS[resource_type.to_s.underscore.camelcase] || Bolognese::Utils::DC_TO_RIS_TRANSLATIONS[resource_type_general.to_s.dasherize] || "GEN"
```
Comment (mfenner) [inline lib/bolognese/readers/datacite_reader.rb:92]: I think we can expand DC_TO_RIS_TRANSLATIONS with new DC content types.

**Review by mfenner — COMMENTED** [review]
Please look into the mapping resourceTypeGeneral DC -> CR. Currently we are doing DC -> SO -> CR (crossref_reader L94-96), this might no longer be necessary. 

Similar with ris_reader (L42-44), bibtex reader (L34-36), and citeproc_reader (L39-41).

**Comment by richardhallett** [comment]
I've updated the mappings as requested, let me know what you think.

**Review by mfenner — APPROVED** [review]
LGTM

**Pull request merged**